midterm-2023-11-06/code/5184-leona-kirijas/script.ts

```diff
@@ -0,0 +1,189 @@
+interface Band {
+    id: number;
+    name: string;
+    genre: string;
+    formed: number;
+    location: string;
+    country: string;
+    members: string[];
+    albums: Album[];
+    
+}
+
+interface Album{
+    name: string;
+    year: number;
+}
+
+type BandSorter = (first: Band, second: Band) => number;
+
+document.addEventListener("DOMContentLoaded", siteCode)
+
+let bands: Band[] = [];
+
+async function siteCode() {
+    const data = await loadData();
+    bands = data.metalBands;
+
+    fillCountries(bands);
+    displayBands(bands);
+
+    const nameSort = document.getElementById("sort-name")!;
+    nameSort.addEventListener("click", sortByName);
+
+    const idSort = document.getElementById("sort-id")!;
+    idSort.addEventListener("click", sortById);
+
+    const yearSort = document.getElementById("sort-formed")!;
+    yearSort.addEventListener("click", sortByYear);
+
+    const locationSort = document.getElementById("sort-Location")!;
+    locationSort.addEventListener("click", sortByLocation);
+
+
+
+    const applyFilterButton = document.getElementById("apply-filter")!;
+    applyFilterButton.addEventListener("click", applyFilter)
+}
+
+
+const nameSorter: BandSorter = (first, second) => first.name.localeCompare(second.name);
+const idSorter: BandSorter = (first, second) => first.id - second.id;
+const yearSorter: BandSorter = (first, second) => first.formed - second.formed;
+const locationSorter: BandSorter = (first, second) => first.name.localeCompare(second.name);
+
+
+const sortByName = () => {
+const sortedBands = bands.toSorted(nameSorter);
+displayBands(sortedBands);
+}
+
+const sortById = () => {
+const sortedBands = bands.toSorted(idSorter);
+displayBands(sortedBands);
+}
+
+const sortByYear = () => {
+    const sortedBands = bands.toSorted(yearSorter);
+    displayBands(sortedBands);
+    }
+const sortByLocation = () => {
+const sortedBands = bands.toSorted(locationSorter);
+displayBands(sortedBands);
+}
+
+
+
+const fillCountries = (bands: Band[]) => {
+    const filter = document.getElementById("country-filter")!;
+
+    const countries = new Set<string>();
+    for (const band of bands) {
+        countries.add(band.country);
+    }
+
+    for (const country of countries) {
+        const option = document.createElement("option");
+        option.value = country;
+        option.innerHTML = country;
+        filter.appendChild(option);
+    }
+}
+const applyFilter = () => {
+    const bandNameElement = document.getElementById("name-filter") as HTMLInputElement;
+    const bandName = bandNameElement.value.toLowerCase();
+
+    const countryElement = document.getElementById("country-filter") as HTMLSelectElement;
+    const country = countryElement.value;
+
+    const genreElement = document.getElementById("genre-filter") as HTMLSelectElement;
+    const genre = bandNameElement.value.toLowerCase();
+    let filteredBands = bands;
+
+    if (bandName) {
```

Comment:
The genre filter is not applied, and the country filter will not work because the bands do not have a country value. +5